### PR TITLE
IDE: When cancelling during typecheck, wait for all workers to exit before exiting typecheck

### PR DESCRIPTION
**Pull request by jvilk-stripe**
IDE: When cancelling during typecheck, wait for all workers to exit before exiting typecheck.

We don't need this logic in resolver or elsewhere, as we don't instrument the multiplexed loops in those passes.

### Motivation

I believe this will fix a segfault that happens sometimes during typechecking. The lambda passed to worker threads captures some variables by reference, and if we don't wait for the lambdas to complete before exiting from the function, the references will no longer be valid -- causing segfaults and such.

The segfault is fairly rare -- likely because, in the common case, the workers exit before the loop hits its 250ms timeout waiting for results on the queue.

### Test plan

There's no easy way to test this as far as I can tell.

### CODE REVIEW

**Comment by DarkDimius** [comment]
Can we somehow have cough this with ENFORCE's?

File: main/pipeline/pipeline.cc [hunk 1]
```diff
@@ -979,6 +982,7 @@ ast::ParsedFilesOrCancelled typecheck(unique_ptr<core::GlobalState> &gs, vector<
                     threadResult.counters = getAndClearThreadCounters();
                     resultq->push(move(threadResult), processedByThread);
                 }
+                workerCounter.DecrementCount();
```
Comment (DarkDimius) [inline main/pipeline/pipeline.cc]: This requires all threads to ever pickup a job.
This means that you're asking for worst case performance of scheduler across cores.
Can we have a scheme that doesn't require that?

Comment (DarkDimius) [reply main/pipeline/pipeline.cc]: Also note that we _pin_ threads to cores. 
If some other process has higher priority on that core, you may wait arbitrarily long.

**Comment by DarkDimius** [comment]
> worker threads captures some variables by reference

can we make it not?

**Comment by jvilk-stripe** [comment]
> can we make it not?

I'm not sure how? GlobalState is one such reference (via `Context`). We could turn GlobalState into a shared pointer, but I think that opens up a whole new can of worms.

**Comment by jvilk-stripe** [comment]
Another alternative: Have threads falsely report that they have completed all of their work units. In other words, report `processedByThread` but don't actually do the work. Then, we can move the cancelation check out of that main loop.

I can put together that change.

**Comment by DarkDimius** [comment]
> I'm not sure how? GlobalState is one such reference (via Context). We could turn GlobalState into a shared pointer, but I think that opens up a whole new can of worms.

these are two separate problems and we need to fix both:
 - we should have a mechanism to ensure that no global states get shared in an unsafe way
 - we shouldn't pass references in lambdas that might outlive function.

**Comment by DarkDimius** [comment]
> Have threads falsely report that they have completed all of their work units. In other words, report processedByThread but don't actually do the work. Then, we can move the cancelation check out of that main loop.

why would this be necessary? code already explicitly checks for cancellation, right?

**Comment by jvilk-stripe** [comment]
> why would this be necessary?

To avoid the race between exiting the thread and exiting the lambda. Then I wouldn't have to make changes to how we pass context to the lambdas.

**Pull request merged**